docs/helpers.md

```diff
@@ -0,0 +1,97 @@
+Helper Structs
+--------------
+
+##### Memo
+
+Linked list, optimized for quickly finding a given type.
+
+- type: int, either a token or a rule (rules start at 256)
+- node: NULL or pointer to AST node OR pointer to token object
+- mark: if node != NULL, index into Parser's array of tokens
+- next: NULL or pointer to next Memo structure
+
+##### Token
+
+These are in an array linked from Parser.  (Or linked list???)
+
+- type: int, token type (needs only 8 bits)
+- bytes: bytes object
+- lineno, col_offset, end_lineno, end_col_offset: int
+- memo: Pointer to linked list Memo
+
+##### Parser
+
+Also needs to point to a PyArena, used for allocating AST nodes and
+other things.  And maybe some flag indicating there's an allocation
+error and a convention to bail if there is one.
```

Comment:
Allocation errors were always dealt with by bailing immediately (Pablo added this).